### ISSUE TITLE
Consumer pings

### ISSUE THREAD

**Issue by davidsteinsland** [bernardphp/bernard#192]
Is it possible to somehow ping the consumers, or let the consumers notify when they are active? Would be very nice to assert that there are consumers running before dispatching work (if not fallback to "normal" approach).

**Comment by henrikbjorn** [comment]
There is nothing currently that allows you to ping consumers. Also the great thing about using a queue is that you dont need to know if a consumer is running when producing messages.

However it is fairly simple to wrap the consumer with code that updates a redis list or a database everytime it checks for a message, then you can check that updated_at field before producing. Juno even have a Controller you can overwrite and show that information on its frontpage.

**Comment by davidsteinsland** [comment]
For non-critical tasks queues are good, but when sending out one-time PIN codes on e-mail, then it's very important that the email is sent -- no matter what. 

I had hoped I wouldn't have to wrap the drivers, as I don't want to bind myself to a single one. A very simple solution would to dispatch an `bernard.present` event or something in the `Consumer::tick()` function. That way it's up to me how I want to store that information, e.g. in a database with list of consumers are last check-in time.

**Comment by henrikbjorn** [comment]
That is a good idea, if you could do a PR for that. Thinking about it, we could have it dispatch the event every 100th tick or something?

I would think a better solution would be to monitor the consumers instead of checking if a consumer is running before dispatching.

**Comment by davidsteinsland** [comment]
I could create a PR, but I see now that I'm using `0.12.2` and master have evolved a bit since then. If you could tag master as a pre-release, I can get around to creating the PR.

**Comment by henrikbjorn** [comment]
You mean an 1.0.0-alpha1 ?

**Comment by davidsteinsland** [comment]
Most likely yes, since it breaks a few things compared to `0.x` releases. Should probably just tag it as `1.0.0-alpha`, then `1.0.0-alpha.1` and so forth.

**Comment by henrikbjorn** [comment]
1.0.0-alpha1 released.

**Comment by davidsteinsland** [comment]
Looking at the code now: what's the reason behind `declare(ticks=1)`? To solve the tick issue simple, I thought of `declare(ticks=100)` and `register_tick_function([$this, 'onTick'])`.

However, since the drivers/queues are sleeping, this will not be accurate: if there is a 5 second delay between each `Consumer::tick()` call, it's easy to get a 30 second delay between each `onTick` call.

**Comment by davidsteinsland** [comment]
Also: why is the drivers/queues sleeping, instead of the consumer?

```php
public function tick(Queue $queue, array $options = []) {
    // ...

    if ($timeSinceLastTick < $this->options['time-between-each-job-check']) {
        return true;
    }

    // ...
}
```

**Comment by henrikbjorn** [comment]
@davidsteinsland not all of the drivers are doing sleep, some of the drivers support "waiting" calls for messages. This is extremely useful for SQS where you pay per request.

I believe the declare(tick=1) was added for making sure we capture signals as soon as possible, instead of waiting for the tick cycle.

**Comment by davidsteinsland** [comment]
Ah, yes. `pcntl` uses ticks to deliver the signals.

What about something like this, in the `Consumer::consume()` method?

```php
        do {
            $now = microtime(true);
            $continue = $this->tick($queue, $options);
            $then = microtime(true);

            if (($then - $now) >= $this->options['ping-delay']) {
                $this->dispatcher->dispatch('bernard.ping');
            }
        } while ($continue);
```

`options['ping-delay']` would then be an unsigned number specifying delay in seconds between each event.

**Comment by henrikbjorn** [comment]
Depending on how real time this needs to be, i think we could just do it every 10 seconds and not add another action :)

**Comment by henrikbjorn** [comment]
Or just do it everytime we tick.

**Comment by ruudk** [comment]
This can be closed since #197 is merged.
